File: module_4/python/first_missing_positive.py
"""
Title: First missing number

Problem: Given an unsorted integer array, find the smallest missing positive
integer.


Execution: python first_missing_positive.py
"""
import unittest
from typing import List


def first_missing_positive(nums: List[int]):
    for i in range(len(nums)):
        if nums[i] <= 0 or nums[i] > len(nums):
            nums[i] = float('inf')

    for i in range(len(nums)):
        absolute_value = abs(nums[i])
        if absolute_value <= len(nums):
            nums[absolute_value-1] = -abs(nums[absolute_value-1])

    for i in range(len(nums)):
        if nums[i] > 0:
            return i+1

    return len(nums)+1

# A slightly different implementation
def first_missing_positive_alt(nums: List[int]):
    """First missing positive position."""
    nums.append(0)
    n = len(nums)

    for i in range(len(nums)):
        # Delete useless elements.
        if nums[i] < 0 or nums[i] >= n:
            nums[i] = 0
    for i in range(len(nums)):
        # Use the index as the hash to record frequency of each number.
        nums[nums[i] % n] += n
    for i in range(1, len(nums)):
        if nums[i]/n == 0:
            return i
    return n


class TestFirstMissingPositive(unittest.TestCase):
    """Unit test for first_missing_positive."""

    def test_1(self):
        self.assertEqual(first_missing_positive([1, 2, 0]), 3)

    def test_2(self):
        self.assertEqual(first_missing_positive([3, 4, -1, 1]), 2)

    def test_3(self):
        self.assertEqual(first_missing_positive([7, 8, 9, 11, 12]), 1)


if __name__ == '__main__':
    unittest.main()

File: module_4/java/FirstMissingPositive.java
/*
 *   Title: First missing positive.
 *
 *   Problem: Given an unsorted integer array, find the smallest missing
 *   positive integer.
 *
 *   Execution: javac FirstMissingPositive.java && java FirstMissingPositive
 */
import java.util.*;


public class FirstMissingPositive {

    public static int firstMissingPositive(int[] A) {
        for (int i = 0; i < A.length; i++) {
            if (A[i] <= 0 || A[i] > A.length) A[i] = Integer.MAX_VALUE;
        }

        for (int i = 0; i < A.length; i++) {
            int abs = Math.abs(A[i]);
            if (abs <= A.length) A[abs-1] = -Math.abs(A[abs-1]);
        }

        for (int i = 0; i < A.length; i++) {
            if (A[i] > 0) return i+1;
        }

        return A.length+1;
    }

    // Alternate implementation
    public static int firstMissingPositiveAlt(int[] A) {
        int i = 0;
        while(i < A.length){
            if(A[i] == i+1 || A[i] <= 0 || A[i] > A.length) i++;
            else if(A[A[i]-1] != A[i]) swap(A, i, A[i]-1);
            else i++;
        }
        i = 0;
        while(i < A.length && A[i] == i+1) i++;
        return i+1;
    }

    private static void swap(int[] A, int i, int j){
        int temp = A[i];
        A[i] = A[j];
        A[j] = temp;
    }

    public static void main(String[] args) {
        int[] test_input_1 = {1, 2, 0};
        assert firstMissingPositive(test_input_1) == 3;

        int[] test_input_2 = {3, 4, -1, 1};
        assert firstMissingPositive(test_input_2) == 2;

        int[] test_input_3 = {7, 8, 9, 11, 12};
        assert firstMissingPositive(test_input_3) == 1;

        System.out.println("Passed all test cases");
    }

}
